src/core/server/saved_objects/migrationsv2/next.ts

```diff
@@ -63,19 +61,28 @@ export const nextActionMap = (client: ElasticsearchClient, transformRawDocs: Tra
       Actions.createIndex(client, state.targetIndex, state.targetIndexMappings),
     CREATE_REINDEX_TEMP: (state: CreateReindexTempState) =>
       Actions.createIndex(client, state.tempIndex, state.tempIndexMappings),
-    REINDEX_SOURCE_TO_TEMP: (state: ReindexSourceToTempState) =>
-      Actions.reindex(
+    REINDEX_SOURCE_TO_TEMP_OPEN_PIT: (state: ReindexSourceToTempOpenPit) =>
+      Actions.openPit(client, state.sourceIndex.value),
+    REINDEX_SOURCE_TO_TEMP_READ: (state: ReindexSourceToTempRead) =>
+      Actions.readWithPit(
         client,
-        state.sourceIndex.value,
+        state.sourceIndexPitId,
+        state.unusedTypesQuery,
+        state.batchSize,
+        state.lastHitSortValue
+      ),
+    REINDEX_SOURCE_TO_TEMP_CLOSE_PIT: (state: ReindexSourceToTempClosePit) =>
+      Actions.closePit(client, state.sourceIndexPitId),
+    REINDEX_SOURCE_TO_TEMP_INDEX: (state: ReindexSourceToTempIndex) =>
+      Actions.transformDocs(
```

Comment:
Why are we transforming documents twice? We run through the transforms when we're reindexing the docs batch by batch from the source index to the temp index and then again in the [temp index](https://github.com/elastic/kibana/pull/97222/files?file-filters%5B%5D=.json&file-filters%5B%5D=.ts#diff-21294d43c3f26d9f80368e326b6cec4050c90f4d71c9d6a61a46141ddd441d56R107). The difference is that when we run through the transforms for docs in the temp index, we're not using a `PIT` search.

Reply:
Hum that seems very right. We are transforming documents in `REINDEX_SOURCE_TO_TEMP_INDEX`, during the `source`->`temp` PIT-based copy

and then later once the docs are in the `target` index, in `OUTDATED_DOCUMENTS_TRANSFORM`.

I understand we had to switch from a plain `reindex` of the `source` -> `temp` indexes because of this ID rewriting, but then (once again, if my understanding is correct) we are effectively migrating the documents (like all the documents, not only the outdated ones) during the `REINDEX_SOURCE_TO_TEMP_INDEX` operation.

I also think that version-based migration are a no-op if a document is already at the latest version, so having all documents passing though `transformDocs` during `REINDEX_SOURCE_TO_TEMP_INDEX` is probably alright.

If that's correct, I guess the outdated documents are effectively migrated before the `CLONE_TEMP_TO_TARGET` operation, right? Would the solution be to just remove the `OUTDATED_DOCUMENTS_SEARCH` and `OUTDATED_DOCUMENTS_TRANSFORM` operations then?

Reply:
Looking at [this comment](https://github.com/elastic/kibana/pull/93002/files#r588269456), it seems the `OUTDATED_DOCUMENTS_SEARCH` is still necessary.

However I'm not sure to understand why.

> On this step, we transform outdated docs that haven't been processed on REINDEX_SOURCE_TO_TEMP_INDEX

How can documents not be processed in `REINDEX_SOURCE_TO_TEMP_INDEX`? Aren't we processing all documents in that step?

Reply:
Ok, I need to RTFM I guess

> If another instance has a disabled plugin it will reindex that plugin's
documents without transforming them. Because this instance doesn't know which
plugins were disabled by the instance that performed the
`REINDEX_SOURCE_TO_TEMP_INDEX` step, we need to search for outdated documents
and transform them to ensure that everything is up to date.

Reply:
@pgayvallet pointed out at an edge case in `OUTDATED_DOCUMENTS_SEARCH` step: as Kibana migrates docs in a single index, the index still keeps the old documents after migration, which may cause data loss during the next migration due to an old object rewriting an already migrated document. looking at the problem.

Reply:
To avoid that we can add a cleanup step for `OUTDATED_DOCUMENTS_SEARCH` phase to remove documents migrated during `OUTDATED_DOCUMENTS_SEARCH` step.
iterate over an array of `outdatedDocs` passed to `OUTDATED_DOCUMENTS_TRANSFORM` step, if an object with a given id doesn't exist in `output` array, add it to the list of objects for delition.
```js
// OUTDATED_DOCUMENTS_TRANSFORM: (input) => output
const elementsToDelete = input.filter((input_element) => output.every((output_element) => output_element.id !== input_element.id));
es.delete(elementsToDelete);
```

Reply:
> we can add a cleanup step for OUTDATED_DOCUMENTS_SEARCH phase to remove documents migrated during OUTDATED_DOCUMENTS_SEARCH step.

@mshustov that might help. I think we do, however, need a guarantee that we don't end up trying to transform docs that _failed_ the transform step in a previous batch returned from `search` by using the same pitSearch you've used previously. I know that at the moment, we fail immediately on a transformation error but the [follow up](https://github.com/elastic/kibana/pull/96986) to that won't fail immediately. We're going to end up in a loop of trying to transform docs that we already couldn't.

Neither @pgayvallet nor I know if that should be handled in this PR or in the [follow up](https://github.com/elastic/kibana/pull/96986).

Reply:
@TinaHeiligers, let me clarify: we need to modify the `OUTDATED_DOCUMENTS_SEARCH` step to use PIT in `search` to make Kibana doesn't try to migrate the same SO indefinitely. Does it sound correct to you? If so, I don't think it's related to the current PR because we have to address the same problem for the `master` implementation as well. I can help you with it in a follow-up PR.

Reply:
> we need to modify the OUTDATED_DOCUMENTS_SEARCH step to use PIT in search to make Kibana doesn't try to migrate the same SO indefinitely

That's correct

> If so, I don't think it's related to the current PR because we have to address the same problem for the master

Doesn't seems necessary for current PR, but is for https://github.com/elastic/kibana/pull/96986, so it probably make sense to have it done in https://github.com/elastic/kibana/pull/96986 directly, wdyt @TinaHeiligers ?

Reply:
created https://github.com/elastic/kibana/issues/97965. we can add PIT there or in #96986. @TinaHeiligers

Reply:
@pgayvallet @mshustov I'll add it to https://github.com/elastic/kibana/pull/96986 if it hasn't been done in https://github.com/elastic/kibana/issues/97965. Technically, it's a requirement for the implementation in https://github.com/elastic/kibana/pull/96986 and makes sense to do it there.
To prevent duplicating work, I might wait for this PR to be merged and take it from there. @mshustov I am hoping to reuse the `Actions.openPit(...)` and `Actions.readWithPit(...)`. Would that make sense?

Reply:
@TinaHeiligers I don't see why it wouldn't work.

Reply:
I'll get to a detailed review later today but definitely tomorrow before going on PTO.